### PR TITLE
Improve performance to splitting logs in FilterEvents

### LINKED ISSUE #19406: Optimize the sample log splitting in FilterEvents
### Expected behavior

FilterEvents shall clone the non-TimeSeriesProperty from the parent workspace (i.e., the input EventWorkspace).  But for the TimeSeriesProperty sample logs, the child workspaces should not have these logs until they are split.  

Furthermore, FilterEvents should allow user to exclude arbitrary sample log from cloning or split in the case for some TimeSeriesProperty log with huge size.

### Actual behavior

FilterEvents clones all the properties to the child workspaces regardless whether they will be split or not.

### PR DESCRIPTION
**Pull request by wdzhou**
1. Use createWithoutLogs() to create child workspaces. Copy the non-time-series and single-value time-series sample logs to child workspaces. Split the time series logs and add to child workspaces.

2. Fix a bug: raise an exception if the base name of output workspace is same as input workspace and GroupWorkspace is set to True.

3. Do not set output properties for each output workspace if GroupWorkspace is set to True.

**To test:**

1. check the new unit tests;
2. filter events by excluding some sample logs from splitting and copying to child workspaces.

Fixes #19406. 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by wdzhou** [comment]
Failed system test.

CompareMDWorkspaces started
The workspaces did not match: MDHistoWorkspaces have a different signal at index 2067 (0.00173095 vs 0.00173125)
CompareMDWorkspaces successful, Duration 0.00 seconds

It is due to the minor change in log splitting, which is acceptable.

**Comment by peterfpeterson** [comment]
This current version breaks `SNSPowderReduction` which will need to be modified before the PR is merged.
```python
Load(
                       Filename = 'NOM_78241.nxs.h5',
                       OutputWorkspace = 'NOM_78241_meta',
                       MetaDataOnly = '1')

# Construct the event filters
GenerateEventsFilter(
                       InputWorkspace  = 'NOM_78241_meta',
                       OutputWorkspace = 'NOM_78241_splitters',
                       InformationWorkspace = 'NOM_78241_splitinfo',
                       TimeInterval   = '60.0',
                       UnitOfTime = 'Seconds',
                       LogName    = '',
)
SNSPowderReduction(
                       SaveAs = 'gsas',
                       FinalDataUnits = 'MomentumTransfer',
                       ExpIniFilename = '/SNS/NOM/IPTS-17210/shared/autoNOM_standards_Aug12/exp.ini',
                       Filename = 'NOM78241',
                       Sum = '0',
                       BinInDspace = '1',
                       Binning = '-0.0001000',
                       OutputDirectory = '/tmp',
                       CalibrationFile = '/SNS/users/pf9/NOM_d44702_2015_06_25.h5',
                       CharacterizationRunsFile = '/SNS/NOM/shared/CALIBRATION/2016_2_1B_CAL/NOM_char_2016_12_13-pdf.txt',
                       FilterBadPulses = '25.0',
                       PreserveEvents = '1',
                       PushDataPositive = 'None',
                       RemovePromptPulseWidth = '50.0',
                       StripVanadiumPeaks = '1',
                       SplittersWorkspace = 'NOM_78241_splitters',
                       SplitInformationWorkspace='NOM_78241_splitinfo',
                       )
```

**Review by peterfpeterson — APPROVED** [review]
Ready to go now

File: Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h [hunk 1]
```diff
@@ -74,6 +74,8 @@ class DLLExport FilterEvents : public API::Algorithm {
   }
 
 private:
+  friend struct Mantid::Kernel::CreateUsingNew<FilterEvents>;
```
Comment (martyngigg) [inline Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h]: Is this used for anything? `CreateUsingNew` is generally used by the `SingletonHolder` classes. I can't see why it is needed here.

File: Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h [hunk 2]
```diff
@@ -127,6 +132,49 @@ class DLLExport FilterEvents : public API::Algorithm {
...
+      std::vector<Kernel::TimeSeriesProperty<int> *> &split_tsp_vec);
+
+  void generateSplitterTSPalpha(
+      std::vector<Kernel::TimeSeriesProperty<int> *> &split_tsp_vec);
+
+  ///
+  void mapSplitterTSPtoWorkspaces(
+      const std::vector<Kernel::TimeSeriesProperty<int> *> &split_tsp_vec);
+
+  //  void splitTSPLogInPlace(DataObjects::EventWorkspace_sptr eventws,
```
Comment (martyngigg) [inline Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h]: Dead code. Can we remove it?

File: Framework/Algorithms/src/GenerateEventsFilter.cpp [hunk 3]
```diff
@@ -887,9 +884,21 @@ void GenerateEventsFilter::makeFilterBySingleValue(
...
     else
       stop = currT;
-    addNewTimeFilterSplitter(start, stop, wsindex, info);
-    numgood = 0;
+
+    std::string empty("");
+    addNewTimeFilterSplitter(start, stop, wsindex, empty);
   }
+
+  // add the row for information table workspace
```
Comment (martyngigg) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]: Dead code. Can we clean this up too?

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py [hunk 4]
```diff
@@ -1474,6 +1474,7 @@ def _split_workspace(self, raw_ws_name, split_ws_name):
                              SplitterWorkspace=split_ws_name, GroupWorkspaces=True, RelativeTime=is_relative_time)
         else:
             # split with information table
+            self.log().notice('[DB...BAT] Raw Name: {0}.  Base Name: {1}'.format(raw_ws_name, base_name))
```
Comment (martyngigg) [inline Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py]: Does this belong at `notice` or `information` level?

File: Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h (hunk 1 above)
Comment (wdzhou) [inline Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h]: This is not useful at all. It was from a failed attempt.  But I forgot to delete.

File: Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h (hunk 2 above)
Comment (wdzhou) [inline Framework/Algorithms/inc/MantidAlgorithms/FilterEvents.h]: mapSplitterTSPtoWorkspaces() is used.

I removed line 156.

File: Framework/Algorithms/src/GenerateEventsFilter.cpp (hunk 3 above)
Comment (wdzhou) [inline Framework/Algorithms/src/GenerateEventsFilter.cpp]: Done!

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py (hunk 4 above)
Comment (wdzhou) [inline Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py]: It was for debugging.  I forgot to delete it afterwards.  Meanwhile, I cleaned up some other debugging outputs assigned to wrong logging level.  Thanks!

**Review by martyngigg — APPROVED** [review]
Thanks, @wdzhou. Looks good.

**Pull request merged**